File: project_euler_python/problem_4.py
'''
Project Euler

Below is the solution to problem #4

A palindromic number reads the same both ways. The largest palindrome made
from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.


'''
import math

def is_palindrome(string):
    length = len(string)/2
    count = 0
    while count <= length:
        if string[count] != string[-1 - count]:
            return False
        count += 1
    return True

def exhaustive(num_digits):
    min = int(math.pow(10, num_digits-1))
    max = int(math.pow(10, num_digits)-1)
    x = min
    y = min
    palindromes = []
    while x < max:
        while y < max:
            val = x * y
            if is_palindrome(str(val)):
                palindromes.append(val)
            y += 1
        x +=1
        y = x
    return palindromes
    


if __name__=="__main__":
    p = exhaustive(3)
    print max(p)
    

File: project_euler_javascript/problem_4.js
/*
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

var isPalindrome = function(str){
  var start = 0;
  var finish = str.length-1;
  while(start <= finish){
    if(str[start] !== str[finish]){
      return false;
    }
    start++;
    finish--;
  }
  return true;
};

var allProducts = function(min, max){
  var result = [];
  for ( var i = min; i <= max; i++ ){
    for ( var j = i; j <= max; j++){
      result.push(i*j);
    }
  }
  return result;
};


var largestPalindrome = function(arr){
  var largest = -1;
  for( var i = 0; i < arr.length; i++ ){
    if (arr[i] > largest){
      if (isPalindrome(arr[i].toString())){
        largest = arr[i];
      }
    }
  }
  return largest;
};

console.log(largestPalindrome(allProducts(100, 999)));